lib/datafile.js

```diff
@@ -403,32 +399,28 @@ _.extend(Reader.prototype, {
                 this.decoder
         );
 
-        if (block instanceof AvroErrors.BlockDelayReadError) {
-            cb(block);
-        } else {
-            // If the sync marker doesn't match, maybe there is no sync
-            // marker, try skipping back
-            if (block.sync && block.sync.toString() !== this.header.sync.toString()) {
-                self._fileBlock.skip(-SYNC_SIZE);
-            }
+        // If the sync marker doesn't match, maybe there is no sync
+        // marker, try skipping back
+        if (block.sync && block.sync.toString() !== this.header.sync.toString()) {
+            self._fileBlock.skip(-SYNC_SIZE);
+        }
 
-            var codec = this.header.meta['avro.codec'].toString();
-            this.decompressData(block.objects, codec, function(err, data) {
-                if (err) {
-                    self.emit('error',  err);
-                } else {
-                    if (data) {
-                        self._datumBlock.write(data);
-                        self.decoder.input(self._datumBlock);
-                        while (block.objectCount--) {
-                            var decoded = self.datumReader.read(self.decoder);
-                            if (decoded !== null) self.push(decoded);
-                        }
-                        cb();
+        var codec = this.header.meta['avro.codec'].toString();
+        this.decompressData(block.objects, codec, function(err, data) {
+            if (err) {
+                self.emit('error',  err);
+            } else {
+                if (data) {
+                    self._datumBlock.write(data);
+                    self.decoder.input(self._datumBlock);
+                    while (block.objectCount--) {
+                        var decoded = self.datumReader.read(self.decoder);
+                        if (decoded !== null) self.push(decoded);
                     }
+                    cb();
                 }
-            });
-        }
+            }
+        });
     },
 
     _readBlocks: function(cb) {
```

Comment:
Yes, the callback breaks the recursion because we don't have enough data.  When more data are available, _transform will be called by the framework, leading to more data being written to the Block and _readBlocks being called again.